Add plugin hooks information

toolmantim opened the pull request:
This reworks our hooks page to include information about plugin hooks.

The "Available hooks" have been redone to be in a table just like environment variables. Hopefully this makes it much easier to read and scan, and explicitly states the order that the hooks are run in.

<img width="748" alt="list-of-hooks" src="https://user-images.githubusercontent.com/153/40107358-5a67ac40-593b-11e8-9ec4-391d4ed0557e.png">

It also updates the plugin doc to link back to the hooks page in lots of places, because it's important for plugin authors to understand hooks, and how they're run in order and relation to all the other hooks.

You might need to check this one out locally @harrietgrace.

toolmantim commented:
I just updated this to the latest master. No content changes though.

harrietgrace commented:
Mmm, I guess they are 'repository' level rather than pipeline... if you have multiple pipelines for one respository?

toolmantim commented:
Yeah! And they're checked into your repository. I'm thinking perhaps they should be called "Repository hooks" now…

toolmantim commented:
Okay, I've updated it to repository hooks:

<img width="853" alt="image" src="https://user-images.githubusercontent.com/153/40344894-3cab774c-5dda-11e8-9ea6-47d9b8278fe3.png">

and updated the table:

<img width="896" alt="new-hooks" src="https://user-images.githubusercontent.com/153/40344886-30ca4f7a-5dda-11e8-900e-46016ed3fac4.png">

One thing I've also done though, is that I removed the "running in a container" from the command hook description but changed it from "script running behaviour" to "command running". I think most people's container usage would be via the docker/compose plugins, and that there's lots of plugins that override the command for other purposes (like how the [detect-clown plugin does](https://github.com/buildkite-plugins/detect-clowns-buildkite-plugin/blob/master/hooks/command)). It feels weird not having what it's useful for tho! Maybe we can re-work it?

harrietgrace commented on pages/agent/v3/hooks.md.erb:
```diff
@@ -1,40 +1,40 @@
+  <tbody>
+    <tr><td style="white-space: nowrap"><code>environment</code></td><td style="white-space: nowrap">Agent, Plugin</td><td>Runs before all other commands. Useful for <a href="/docs/agent/v3/securing#using-environment-hooks-for-secrets">exposing secret keys</a> and <a href="/docs/agent/v3/securing#whitelisting">security whitelisting</a>.</td></tr>
+    <tr><td style="white-space: nowrap"><code>pre-checkout</code></td><td style="white-space: nowrap">Agent, Plugin</td><td>Runs before checkout.</td></tr>
+    <tr><td style="white-space: nowrap"><code>checkout</code></td><td style="white-space: nowrap">Agent, Plugin</td><td>Overrides the default <code>git checkout</code> behavior. <em>Note:</em> If multiple checkout hooks are found, only the first will be run.</td></tr>
```
Should checkout be around the other way? Re: the chats in campfire just now

toolmantim replied:
Doh! Yes, yes it should! Update :) Thanks for the catch.

toolmantim commented:
Yep! I'm 99% sure. Merge party!!
